### PR TITLE
Advance deprecation of delete expressions, class allocators, class deallocators: Error

### PR DESCRIPTION
**Pull request by JinShil**
Documentation update: https://github.com/dlang/dlang.org/pull/2631

### CODE REVIEW

**Comment by wilzbach** [comment]
```
home/braddr/sandbox/at-client/pull-3607611-Linux_32_64/dmd/generated/linux/release/32/dmd -conf= -m64 -Irunnable  -fPIC  -odgenerated/runnable -ofgenerated/runnable/test23_0  runnable/test23.d 
runnable/test23.d(709): Deprecation: class allocators have been deprecated, consider moving the allocation strategy outside of the class
runnable/test23.d(717): Deprecation: class deallocators have been deprecated, consider moving the deallocation strategy outside of the class
runnable/test23.d(741): Deprecation: class allocators have been deprecated, consider moving the allocation strategy outside of the class
runnable/test23.d(749): Deprecation: class deallocators have been deprecated, consider moving the deallocation strategy outside of the class
runnable/test23.d(728): Error: The `delete` keyword has been deprecated.  Use `object.destroy()` (and `core.memory.GC.free()` if applicable) instead.


==============================
Test runnable/test23.d failed: expected rc == 0, exited with rc == 1

Makefile:153: recipe for target 'generated/runnable/test23.d.out' failed
make[2]: *** [generated/runnable/test23.d.out] Error 1
```

There's also at least one test in druntime that needs to be disabled before this can pass.

**Comment by JinShil** [comment]
> There's also at least one test in druntime that needs to be disabled before this can pass.

Please elaborate.  I replaced `delete` with `destroy` and it passes (though there are other tests that don't pass with such a fix).

**Comment by wilzbach** [comment]
I'm talking about the unittests in here: https://github.com/dlang/druntime/blob/master/src/rt/lifetime.d which test the implementation of delete.

I think they need to be disabled first for this PR to pass.

**Comment by JinShil** [comment]
> I think they need to be disabled first for this PR to pass.

Ok, I'll get to it.  It looks like this is going to be a much bigger job than I anticipated.  It'd probably be a good idea to mark it WIP until I work through all of the issues.

**Comment by rainers** [comment]
I'm not a big fan of deprecation or removal of `delete`. As far as I know the rationale was that it looks innocent syntactically, but is unsafe. The same can be said about other constructs of the language, e.g. pointer offsetting.

On the other hand, it is quite inconvenient to get the call to GC.free correct. Even experienced D programmers don't know how to do it (see e.g. https://issues.dlang.org/show_bug.cgi?id=13558, also slightly related: https://issues.dlang.org/show_bug.cgi?id=19539). You should not have to know to begin with, as it will lock the user code with the internal implementation.

This can be solved by a library function that knows about all the special cases, but I think it should exist before removal. The pair of destroy and delete library functions is unlikely to be accessible to the compiler, while `delete` has defined semantics. With the help of escape analysis it might even proove its usage to be safe.

tldr: If `delete` is removed, a library function should be provided that actually frees the memory. I'd prefer de-deprecation, though.

**Comment by wilzbach** [comment]
@rainers: there's __delete in core.memory which is supposed to be a 1:1 replacement of delete.

**Comment by rainers** [comment]
> @rainers: there's __delete in core.memory which is supposed to be a 1:1 replacement of delete.

Thanks, I didn't know these. And the implementation is just as wrong as in the bug report given above.

**Comment by JinShil** [comment]
Many of the tests were using `delete` to test class deallocators, so I had to advance the deprecation of `delete`, class allocators, and class deallocators simultaneously to make this work and pass the test suite.

**Comment by JinShil** [comment]
> ../../druntime/import/object.d(642): Error: `rt_finalize` cannot be interpreted at compile time, because it has no available source code
Error: expression `<error>` is not constant
runnable/interpret.d(3207):        while evaluating: `static assert(test6907())`

I don't know if it's possible to implement `__delete` at compile-time.  The following code illustrates the problem.

```D
class C { ~this() { ++dtor1; } }
{ Object o = new C(); __delete(o); assert(dtor1 == 4); }
```
Since `__delete` is called with a type of `Object`, the following if statement will return `false` because `T` is `Object` instead of `C`:

```D
static if (__traits(hasMember, T, "__xdtor"))
     obj.__xdtor();
```

`typeid()` isn't implemented for CTFE and CTFE can't perform casts like `ClassInfo** c = cast(ClassInfo**)(&obj)`.  So, I'm stuck.  I don't know how to get the destructor of `C` when `C` is given to me as `Object`.

**Comment by jacob-carlborg** [comment]
> typeid() isn't implemented for CTFE

You can always try to implement that 😉.

**Comment by JinShil** [comment]
I'm closing this for now.  Removing `delete` is going to require a significant amount of work.
